pages/watch/[id].tsx

```diff
@@ -0,0 +1,151 @@
+import React, { useEffect, useState } from "react";
+import { Nav } from "../../components";
+import { useAddress } from "@thirdweb-dev/react";
+import { useRouter } from "next/router";
+import { Player, usePlaybackInfo } from "@livepeer/react";
+import axios from "axios";
+import { ethers } from "ethers";
+import { chains } from "../../constants";
+import { Requirement, TokenInfo } from "../../types";
+
+interface PlaybackPolicy {
+  webhookContext: {
+    requirement: Requirement;
+  };
+}
+
+const Watch: React.FC = () => {
+  const userAddress = useAddress();
+  const [accessStatus, setAccessStatus] = useState<string | boolean>(
+    "checking"
+  );
+  const [jwt, setJwt] = useState<string>("");
+
+  const router = useRouter();
+  const playbackId = router.query.id?.toString();
+
+  const { data: playbackInfo } = usePlaybackInfo({
+    playbackId,
+  });
+
+  const checkAccess = async (playbackPolicy: PlaybackPolicy) => {
+    if (!userAddress) {
+      return;
+    }
+
+    const { requirement } = playbackPolicy.webhookContext;
+    const { token, userAddress: requiredUserAddress } = requirement;
+
+    const hasTokenAndUser =
+      Object.keys(token).length > 0 && requiredUserAddress;
+
+    if (hasTokenAndUser) {
+      const canWatch = await hasSufficientTokenBalance(token);
+      const hasAccess = canWatch && hasWalletAccess(requiredUserAddress);
+      setAccessStatus(hasAccess);
+      if (hasAccess) {
+        await generateJwt(requirement);
+      }
+    } else {
+      if (Object.keys(token).length > 0) {
+        const canWatch = await hasSufficientTokenBalance(token);
+        setAccessStatus(canWatch);
+        if (canWatch) {
+          await generateJwt(requirement);
+        }
+      }
+
+      if (requiredUserAddress) {
+        const canWatch = hasWalletAccess(requiredUserAddress);
+        setAccessStatus(canWatch);
+        if (canWatch) {
+          await generateJwt(requirement);
+        }
+      }
+    }
+  };
+
+  const generateJwt = async (requirement: Requirement) => {
+    const { data } = await axios.post("/api/generate-jwt", {
+      requirement,
+      userAddress,
+    });
+
+    if (data?.token) {
+      setJwt(data.token);
+      setAccessStatus(true);
+    }
+  };
+
+  const hasSufficientTokenBalance = async (tokenInfo: TokenInfo) => {
```

Comment:
biggest blocker, you can use thirdweb to get the balance of any erc20 token without having to provide the ABI if the contract is imported, in case it's not and you want to make sure, you can still pass the ABI to it

https://portal.thirdweb.com/typescript/sdk.thirdwebsdk.smartcontract#abi-optional

basically you should never use ethers directly

Reply:
no problem, i will replace it! thanks

should i import `getContract` from `@thirdweb-dev/react` or other package?

Reply:
getContract from `@thirdweb-dev/sdk@` or useContract from react, for here you can do useContract and pass the ABI as the second param

Reply:
Thanks, I tried it:

```
  const { contract } = useContract(
    requirement?.token?.tokenAddress,
    minABI
  );

  const { data, isLoading, error } = useContractRead(contract, "balanceOf", [
    useAddress
  ]);
  ```
  
But it is showing this error: 

```
Error: The address you're trying to send a transaction to is not a smart contract. Make sure you are on the correct network and the contract address is correct
```

Both contract address and user address is valid

Reply:
> Thanks, I tried it:
> 
> ```
>   const { contract } = useContract(
>     requirement?.token?.tokenAddress,
>     minABI
>   );
> 
>   const { data, isLoading, error } = useContractRead(contract, "balanceOf", [
>     useAddress
>   ]);
> ```
> 
> But it is showing this error:
> 
> ```
> Error: The address you're trying to send a transaction to is not a smart contract. Make sure you are on the correct network and the contract address is correct
> ```
> 
> Both contract address and user address is valid

Regards @suhailkakar , do you have the contract address and in which chain you were connected?

Noticed that there is a type in the code.

```js
  const { contract } = useContract(
    requirement?.token?.tokenAddress,
    minABI
  );

  const { data, isLoading, error } = useContractRead(contract, "balanceOf", [
    useAddress // Tpyo: it should be userAddres
  ]);
  ```
  
actually `useAddress` is the name of the hook.

Can I help addressing Nacho's comments? :pray: 

Regards